tests/testsuite/feature_unification.rs

```diff
@@ -0,0 +1,112 @@
+
+use cargo_test_support::prelude::*;
+use cargo_test_support::{project, str};
+
+#[cargo_test]
+fn feature_unification() {
+    let p = project()
+        .file(
+            ".cargo/config.toml",
+            r#"
+                [resolver]
+                feature-unification = "workspace"
+            "#,
+        )
+        .file(
+            "Cargo.toml",
+            r#"
+                [workspace]
+                resolver = "2"
+                members = ["common", "ab", "bc"]
+            "#,
+        )
+        .file(
+            "common/Cargo.toml",
+            r#"
+                [package]
+                name = "common"
+                version = "0.1.0"
+                edition = "2021"
+
+                [features]
+                a = []
+                b = []
+                c = []
+            "#,
+        )
+        .file(
+            "common/src/lib.rs",
+            r#"
+                pub fn features() {
+                    #[cfg(feature = "a")]
+                    print!("a");
+                    #[cfg(feature = "b")]
+                    print!("b");
+                    #[cfg(feature = "c")]
+                    print!("c");
+                }
+            "#,
+        )
+        .file(
+            "ab/Cargo.toml",
+            r#"
+                [package]
+                name = "ab"
+                version = "0.1.0"
+                edition = "2021"
+
+                [dependencies]
+                common = { path = "../common", features = ["a", "b"] }
+            "#,
+        )
+        .file("ab/src/main.rs", "fn main() { common::features(); }")
+        .file(
+            "bc/Cargo.toml",
+            r#"
+                [package]
+                name = "bc"
+                version = "0.1.0"
+                edition = "2021"
+
+                [dependencies]
+                common = { path = "../common", features = ["b", "c"] }
+            "#,
+        )
+        .file("bc/src/main.rs", "fn main() { common::features(); }")
+        .build();
+
+    p.cargo("build -p common")
+        .with_stderr_data(str![[r#"
+[COMPILING] common v0.1.0 ([ROOT]/foo/common)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    // `common` should not be rebuilt when building `ab` or `bc`
+    p.cargo("build -p ab")
+        .with_stderr_data(str![[r#"
+[COMPILING] ab v0.1.0 ([ROOT]/foo/ab)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+    p.cargo("build -p bc")
+        .with_stderr_data(str![[r#"
+[COMPILING] bc v0.1.0 ([ROOT]/foo/bc)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.cargo("build --workspace")
+        .with_stderr_data(str![[r#"
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]])
+        .run();
+
+    p.process(&p.bin("ab")).with_stdout_data(str!["abc"]).run();
```

Comment:
This is quite a brilliant way to check the unification result!

We usually use `--verbose` cargo execution and verify flags passed down to rustc. Maybe with that we

* Can replace `main.rs` with `lib.rs`. Building binaries requires linking and may slow down tests. 
* Can remove functions in the `common` lib, as we directly assert rustc invocations.

What do you think?

Reply:
We also try to prefer `cargo check` over `cargo build`

However, checking rustc flags has down sides and I feel like its good to avoid when possible 
- A lot of unrelated data to avoid
- We're not testing the end-user effect

Reply:
I'm okay with writing it either way, but what is the consensus? If it's enough to just verify there are no recompilations then `cargo check` alone would be sufficient, however I can't think of any ways to actually verify the activated features without either using `--verbose` or `cargo build`.

Reply:
> emitting a conditional compile_error! if not all the features are enabled

Sounds awesome!